lib/functions/Core.ts

```diff
@@ -12,21 +8,69 @@ type Term = E.TermExpression;
 // Overloaded Functions
 // ----------------------------------------------------------------------------
 
-// Maps argument types on their specific implementation.
-export type OverloadMap = Map<List<ArgumentType>, E.SimpleApplication>;
+type SearchStack = OverloadNode[];
+
+export class OverloadNode {
+  private implementation?: E.SimpleApplication | undefined;
+  private readonly subTrees: Record<string, OverloadNode>;
+  public constructor() {
+    this.implementation = undefined;
+    this.subTrees = {};
+  }
+
+  /**
+   * Adds an overload to the tree structure considering this as the tree's root.
+   * @param argumentTypes a list of ArgumentTypes that would need to be provided in the same order to
+   * get the implementation.
+   * @param func the implementation for this overload.
+   */
+  public addOverload(argumentTypes: ArgumentType[], func: E.SimpleApplication): void {
+    this._addOverload([ ...argumentTypes ], func);
+  }
+
+  private _addOverload(argumentTypes: ArgumentType[], func: E.SimpleApplication): void {
+    const argument = argumentTypes.shift();
+    if (!argument) {
+      this.implementation = func;
+      return;
+    }
+    const str = argument;
+    if (!this.subTrees[str]) {
+      this.subTrees[str] = new OverloadNode();
+    }
+    this.subTrees[str]._addOverload(argumentTypes, func);
+  }
+
+  /**
+   * @param arg term to try and match to possible overloads of this node.
+   * @returns SearchStack a stack with top element the next node that should be asked for implementation or overload.
+   */
+  public getSubTreeWithArg(arg: Term): SearchStack {
+    const matching = Object.entries(this.subTrees).map(([ type, node ]) =>
```

Comment:
This parts needs some cleaning and documentation (any-cast, magic constants, nested ternary, unneeded casts).

Reply:
The magic constants are just for now actually, they would go away with the implementation of autocasts. The casts are necessary I think because sort thinks the type of the array is `(number | OverladNode)[]` when it is in fact `[number, OverladNode]`. For the any cast, I have no idea, it was used like that in the `Typer` object and I just took it from there.

Reply:
I refactored this a bit. I hope it's better now.